### PR TITLE
[WIP] POC for a query language for filtering ps output (#17569)

### PR DESCRIPTION
**Pull request by jawher**
Feature proposal: #17569
Please head to the linked issue for more details on the query language syntax and for a preliminary discussion on this proposal.
### This is just a quick and dirty POC

with the sole goal of having a working example to build the discussion on.
## Modifications
- Add a new top-level package in docker named `query`. This package parses a given query from a string to an AST. The AST can be used to filter containers, images, volumes, networks and any other list of items according to the query. The only requirement is for the item type to implement an interface (`query.Queryable`)
- Make `containers/Container` implement `query.Queryable`
- Add a `--query` flag to `docker ps`
- Modify `engine-api/client/type/ContainerListOptions` to add a `Query` field (in a true quick and dirty fashion, I modified the concerned files in the vendor directory, but a proper PR to the engine-api project will be created if this proposal is accepted)
- Modify `daemon/list.go` to use the provided query instead of psFilters if it is present. The current implementation is, again, a quick and dirty one, and is incomplete, i.e. doesn't handle all psFilters fields among other things.
## Query language

More details in [bateau's project page](https://github.com/jawher/bateau)
### Boolean fields

```
running
```
### Valued fields

e.g. string, []string, int, duration, size, etc.
Syntax: `field operator value`
Where operator is one of:
- `=` for strict equality, `!=` for strict inequality
- `~` for string contains (LIKE), `!~` for the negated version
- `>`, `>=`, `<` and `<=` for numeric comparisons

examples:

```
exit=1
```

```
name=server_1
```

```
name~"server"
```

```
created > 2h
```
### Negation

Any clause can be negated using the `!` operator.
Examples:

```
!running
```

```
!(exit=1) // though this is equivalent to exit!=1
```
### Anding, Oring

2 or more clauses can be combined using anding `&` or bring `|`:

```
running & created > 2d // running containers created more than 2 days ago
```

```
size>300MB & (created > 2M | docker_version~1.5 | docker_version~1.6) 
```

```
!label.arch | label.arch=x86 // containers without an "arch" label or with one set to "x86"
```
## Open questions

Do you think this proposal has a chance of being accepted ? If the answer is yes, I'm willing to rework this PR to handle:
- at least the current supported filters, if not more fields
- generalise it to work on images, volumes and networks (am I forgetting any other item list in docker ?)
### The query syntax:

I reused the code I wrote for [bateau](https://github.com/jawher/bateau).

What is the preferred syntax between:

```
!label.arch | label.arch=x86
size>300MB & (created > 2M | docker_version~1.5 | docker_version~1.6) 
```

and

```
not label.arch and label.arch=x86
size>300MB and (created > 2M and docker_version~1.5 or docker_version~1.6) 
```
### The query parsing complexity

One of the points raised by @duglin in the original issue (#17569) was the cost of adding and maintaining the parsing code.

The current implementation is comprised of a lexer (200 LOC) and a recursive descent parser (~180 LOC).
I also have a [working prototype with a lexer-less parser](https://gist.github.com/jawher/56ed7ec4ab8ea7e9d5e7) (~240LOC), which I believe will be easier to extend with new operators for example than trying to do so with the current lexer.

Also, the syntax as such should not need to be modified frequently, if at all.
A more frequent task would be to add more fields to filter on.
This task can be be accomplished without touching the parsing logic: see `container/query.go` for how the list of fields of a containers are defined.
### Interaction with the current ps filtering mechanism

The `-f`, `--since` and `--before` filters are a subset of the query language of this proposal.

If this proposal is accepted, one possibility would be to deprecate these 3 flags, and make them exclusive with `--query`:

```
Usage:
docker ps ([-f] [--since] [--before]) | --query)
```
### Implementation details

What do you think about how `query.Queryable`  is directly implemented by the concerned types (Container for example) ?
The other possibility would be to created adapter types which would implement the iface in the `daemon` package, e.g. `containerQueryable`, etc.

### CODE REVIEW

**Comment by runcom** [comment]
I like this. I'm just not totally sold by how much code this pr is adding (and needs to be maintained). Can this be thought as an external tool which is piped to docker ps or whatever? It would be great (as I'm thinking about tools like jq)

**Comment by runcom** [comment]
> 2 - is supporting a query language a full-time job? do we want to sign-up for that?

@duglin's second concern in the linked issue made me thought about an external tool (which I think we could add to the doc if anyone needs to have a more powerful query mechanism)

**Comment by jawher** [comment]
@runcom [bateau](https://github.com/jawher/bateau) is such a tool: all it does is connect to a docker daemon, list all containers and filter them according to the provided query.

However, this approach is very slow and inefficient, as it needs one roundtrip to the daemon for the containers list, and one roundtrip per-container to get the full list of fields to be able to filter on them.

Doing this daemon side is much more efficient, especially when filtering containers based on images and ancestors.

A middle ground could maybe be reached by having a son endpoint which returns the list of all containers with all their info (and not just a couple of them like what `containers/json` does).

**Comment by cpuguy83** [comment]
I haven't looked at this, but I know swarm implements some more advanced filtering for it's stuff, may be worthwhile to look into that.

**Comment by jawher** [comment]
The swarm filtering syntax: https://docs.docker.com/swarm/scheduler/filter/

TLDR:
* operators: `==` and `!=`
* values: supports globs and regex, not the case for this proposal, but easily fixable
* no `and`, `or` and `not` operators

**Comment by calavera** [comment]
I don't think we should add this, sorry. I'll be okay if other maintainers really really want to have it.

**Comment by jawher** [comment]
@calavera:
> I don't think we should add this, sorry

Thanks for the reply: the silence was starting to worry me :laughing:

And no worries about rejecting this :) 
 
That's why I implemented this in the quick and dirty fashion, just to have something to show, and to be sure that this is actually wanted before investing more time on it.

**Comment by cpuguy83** [comment]
I think we need better filter syntax (at least to match what swarm has), maybe not as extensive as this is doing.

**Comment by calavera** [comment]
> I think we need better filter syntax (at least to match what swarm has)

there is an agreement between maintainers about wanting to have parity with Swarm. That would make our life much easier. Unfortunately, this PR is much more complex than that.

I'm going to close it now, but feel free to open a new one if you want to port Swarm's syntax on top of the current upstream.

Thanks a lot.
